std/math/emulated/field_assert.go

```diff
@@ -109,23 +120,32 @@ func (f *Field[T]) IsZero(a *Element[T]) frontend.Variable {
 	// then we can ensure in most cases that no overflows happen.
 
 	// as ca is already reduced, then every limb overflow is already 0. Only
-	// every addition adds a bit to the overflow
+	// every addition adds a bit to the overflow.
+	var res0 frontend.Variable
 	totalOverflow := len(ca.Limbs) - 1
 	if totalOverflow > int(f.maxOverflow()) {
 		// the sums of limbs would overflow the native field. Use the first
 		// approach instead.
-		res := f.api.IsZero(ca.Limbs[0])
+		res0 = f.api.IsZero(ca.Limbs[0])
+		for i := 1; i < len(ca.Limbs); i++ {
+			res0 = f.api.Mul(res0, f.api.IsZero(ca.Limbs[i]))
+		}
+	} else {
+		// default case, limbs sum does not overflow the native field
+		limbSum := ca.Limbs[0]
 		for i := 1; i < len(ca.Limbs); i++ {
-			res = f.api.Mul(res, f.api.IsZero(ca.Limbs[i]))
+			limbSum = f.api.Add(limbSum, ca.Limbs[i])
```

Comment:
doesn't this work only if `len(ca.limbs) * max limb value <= max snark scalar field value` ?

Reply:
Exactly - this is the guard `totalOverflow > int(f.maxOverflow())` checks against. Every limb in non-native Element is up to `nbBits+overflow` bits wide, where `nbBits` is constant and `overflow` is tracked through arithmetic operations.

Every time we do an addition between `nbBits+overflow_1` and `nbBits+overflow_2` wide limbs, the resulting limbs are up to `nbBits+max(overflow_1, overflow_2)+1` bits wide. In this method, as the overflow is same for all limbs, then after we add `n` limbs the sum is up to `nbBits+overflow+n-1` bits wide. This is exactly the computed value `totalOverflow` which we compare against maximum permitted overflow computed by `f.maxOverflow()`. So `limbSum` should never overflow native field in this branch.